File: R/03_concordance/naics_concordance_fetch.R
# naics_concordance_fetch.R
# Downloads the three official Census NAICS revision concordances spanning the
# 2005-2025 benchmark window and flattens them into one tidy edge list.
#
# In:   nothing (network)
# Out:  $DATA/naics_concordance/naics_concordance_edges.csv
# Run:  Rscript R/03_concordance/naics_concordance_fetch.R
#
# rp_apply_threshold.R reads that CSV as the authority for which NAICS6 codes
# are comparable across vintages. A Python equivalent with no package
# dependencies is in naics_concordance_fetch.py.

if (!requireNamespace("readxl", quietly = TRUE)) {


  stop("readxl is not installed. Install it with:\n",
       "  install.packages('readxl')\n",
       "or run the Python equivalent instead: python R/03_concordance/naics_concordance_fetch.py")
}
library(readxl)

local({
  for (up in c(".", "..", "../..", "../../..")) {
    f <- file.path(up, "R/config.R")
    if (file.exists(f)) return(source(f))
  }
  stop("Cannot find R/config.R from ", getwd(),
       ". Run from the repo root, or open north-america-panel.Rproj.")
})

BASE    <- "https://www.census.gov/naics/concordances/"
OUT_DIR <- na_data("naics_concordance")

FILES <- list(
  list(vintage = "2007->2012", fname = "2007_to_2012_NAICS.xls"),
  list(vintage = "2012->2017", fname = "2012_to_2017_NAICS.xlsx"),
  list(vintage = "2017->2022", fname = "2017_to_2022_NAICS.xlsx")
)

dir.create(OUT_DIR, recursive = TRUE, showWarnings = FALSE)

  fetch_if_missing <- function(fname) {
  path <- file.path(OUT_DIR, fname)
  if (file.exists(path)) {
    cat("using cached", path, "\n")
    return(path)
  }
  url <- paste0(BASE, fname)
  cat("downloading", url, "\n")
  part <- paste0(path, ".part")
  ok <- tryCatch({
    utils::download.file(url, part, mode = "wb", quiet = TRUE)
    TRUE
  }, error = function(e) { cat("  download failed:", conditionMessage(e), "\n"); FALSE })
  if (!ok || !file.exists(part) || file.size(part) == 0) {
    unlink(part)
    stop("Could not download ", url)
  }
  file.rename(part, path)
  path
}

read_concordance <- function(path) {
  d <- readxl::read_excel(path, sheet = 1, col_names = FALSE,
                          col_types = "text", .name_repair = "minimal")
  d <- as.data.frame(d, stringsAsFactors = FALSE)
  if (ncol(d) < 3) {
    stop("Only ", ncol(d), " columns in ", basename(path),
         " - expected at least 3 (old code, old title, new code).")
  }

  first_col <- trimws(ifelse(is.na(d[[1]]), "", d[[1]]))
  hdr <- which(grepl("NAICS Code$", first_col[seq_len(min(10, nrow(d)))]))
  if (length(hdr) == 0) {
    stop("Could not find the header row in ", basename(path),
         " - layout changed. Expected a first cell ending in 'NAICS Code' ",
         "within the first 10 rows.")
  }
  list(data = d[seq.int(hdr[1] + 1, nrow(d)), , drop = FALSE],
       header_row = hdr[1])
}

clean <- function(x) {
  x <- ifelse(is.na(x), "", as.character(x))
  trimws(gsub("[\r\n]+", " ", x))
}

edges <- data.frame(vintage = character(0), old_code = character(0),
                    old_title = character(0), new_code = character(0),
                    new_title = character(0), code_changed = character(0),
                    stringsAsFactors = FALSE)

for (f in FILES) {
  path <- fetch_if_missing(f$fname)
  got  <- read_concordance(path)
  d    <- got$data

  old_code  <- clean(d[[1]])
  old_title <- clean(d[[2]])
  new_code  <- clean(d[[3]])
  new_title <- if (ncol(d) >= 4) clean(d[[4]]) else rep("", nrow(d))

  keep <- grepl("^[0-9]{6}$", old_code) & grepl("^[0-9]{6}$", new_code)

  add <- data.frame(
    vintage      = rep(f$vintage, sum(keep)),
    old_code     = old_code[keep],
    old_title    = old_title[keep],
    new_code     = new_code[keep],
    new_title    = new_title[keep],
    code_changed = ifelse(old_code[keep] != new_code[keep], "1", "0"),
    stringsAsFactors = FALSE)

  edges <- rbind(edges, add)
  cat(sprintf("  %s: %d six-digit pairs, %d where the code changed\n",
              f$vintage, nrow(add), sum(add$code_changed == "1")))
}

if (nrow(edges) == 0) {
  stop("No edges parsed - refusing to write an empty concordance.")
}

csv_field <- function(x) {
  needs <- grepl('[",\r\n]', x)
  ifelse(needs, paste0('"', gsub('"', '""', x, fixed = TRUE), '"'), x)
}

out_path <- file.path(OUT_DIR, "naics_concordance_edges.csv")
cols <- c("vintage", "old_code", "old_title", "new_code", "new_title",
          "code_changed")

lines <- c(
  paste(cols, collapse = ","),
  do.call(paste, c(lapply(cols, function(k) csv_field(edges[[k]])), sep = ","))
)

con <- file(out_path, open = "wb")
writeLines(lines, con, sep = "\r\n")
close(con)

chg <- edges[edges$code_changed == "1", c("old_code", "new_code")]
chg <- unique(chg)

nodes <- unique(c(chg$old_code, chg$new_code))
parent <- setNames(nodes, nodes)

find <- function(x) {
  while (parent[[x]] != x) x <- parent[[x]]
  x
}
for (i in seq_len(nrow(chg))) {
  a <- find(chg$old_code[i]); b <- find(chg$new_code[i])
  if (a != b) parent[[a]] <- b
}
comps <- length(unique(vapply(nodes, find, character(1))))

cat(sprintf("\nwrote %s\n", out_path))
cat(sprintf("  %d rows; %d distinct changed-code edges; %d official lineages spanning %d codes\n",
            nrow(edges), nrow(chg), comps, length(nodes)))

cat("\n--- tests ---\n")
fails <- 0
check <- function(label, ok, detail = "") {
  if (isTRUE(ok)) cat(sprintf("  PASS  %s\n", label))
  else { cat(sprintf("  FAIL  %s %s\n", label, detail)); fails <<- fails + 1 }
}

check("all three vintages present",
      setequal(unique(edges$vintage),
               vapply(FILES, function(f) f$vintage, character(1))),
      paste("got", paste(unique(edges$vintage), collapse = ", ")))

check("every code on both sides is exactly six digits",
      all(grepl("^[0-9]{6}$", edges$old_code)) &&
        all(grepl("^[0-9]{6}$", edges$new_code)))

check("code_changed agrees with the codes it describes",
      all((edges$code_changed == "1") == (edges$old_code != edges$new_code)))

check("no duplicate (vintage, old_code, new_code) rows",
      !any(duplicated(edges[, c("vintage", "old_code", "new_code")])))

check("no title contains a raw line break",
      !any(grepl("[\r\n]", c(edges$old_title, edges$new_title))))

autos <- unique(c(
  edges$new_code[edges$old_code %in% c("336111", "336112") & edges$code_changed == "1"],
  edges$old_code[edges$new_code %in% c("336110") & edges$code_changed == "1"]))
check("autos: 336111/336112 -> 336110 is in the official tables",
      "336110" %in% autos,
      paste("got", paste(autos, collapse = ", ")))

petro <- unique(edges$new_code[edges$old_code == "211111" & edges$code_changed == "1"])
check("petroleum: 211111 -> 211120 + 211130 is in the official tables",
      all(c("211120", "211130") %in% petro),
      paste("got", paste(petro, collapse = ", ")))

check("tires 326211 never changes code in any vintage",
      !any(edges$code_changed == "1" &
             (edges$old_code == "326211" | edges$new_code == "326211")))

File: R/03_concordance/naics_concordance_fetch.py
"""
naics_concordance_fetch.py

[PROTOCOL Step 6.2] Check the official NAICS concordance tables for splits,
merges or redefinitions.

Downloads the three official Census NAICS revision concordances spanning the
2005-2025 benchmark window and flattens them into one tidy edge list:

    $DATA/naics_concordance/naics_concordance_edges.csv
        vintage, old_code, old_title, new_code, new_title, code_changed

R/04_related_party/rp_apply_threshold.R reads that CSV as the authority for
which NAICS6 codes belong to the same lineage, falling back to its
year-adjacency heuristic only for the Census trade-specific codes the official
tables cannot cover.

naics_concordance_fetch.R is the primary implementation; this is the
equivalent with no package dependencies (the two newer concordances are .xlsx,
which is a zip of XML and readable with the standard library alone, and the
oldest is legacy .xls).

Set NA_PANEL_DATA_DIR to put the output somewhere other than <repo>/data.

Run from the repo root:
    python R/03_concordance/naics_concordance_fetch.py
"""


import collections
import csv
import os
import re
import sys
import urllib.request
import xml.etree.ElementTree as ET
import zipfile

BASE = "https://www.census.gov/naics/concordances/"
OUT_DIR = os.path.join(
    os.environ.get("NA_PANEL_DATA_DIR", "data"), "naics_concordance")

FILES = [
    ("2007->2012", "2007_to_2012_NAICS.xls"),
    ("2012->2017", "2012_to_2017_NAICS.xlsx"),
    ("2017->2022", "2017_to_2022_NAICS.xlsx"),
]

NS = "{http://schemas.openxmlformats.org/spreadsheetml/2006/main}"

def read_xlsx(path):
    """Read the first worksheet of an .xlsx as a list of row-lists.

    An .xlsx file is a zip archive of XML parts, so this needs nothing beyond
    the standard library. Cells carrying text point into a shared-strings
    table rather than holding the text inline, which is why that table is
    read first.
    """
    z = zipfile.ZipFile(path)
    shared = []
    if "xl/sharedStrings.xml" in z.namelist():
        root = ET.fromstring(z.read("xl/sharedStrings.xml"))
        for si in root.findall(NS + "si"):
            shared.append("".join(t.text or "" for t in si.iter(NS + "t")))
    sheets = sorted(n for n in z.namelist()
                    if re.match(r"xl/worksheets/sheet\d+\.xml$", n))
    root = ET.fromstring(z.read(sheets[0]))
    rows = []
    for row in root.iter(NS + "row"):
        cells = {}
        for c in row.findall(NS + "c"):
            col = re.match(r"([A-Z]+)", c.get("r")).group(1)
            v = c.find(NS + "v")
            if v is None:
                inline = c.find(NS + "is")
                txt = ("".join(t.text or "" for t in inline.iter(NS + "t"))
                       if inline is not None else "")
            elif c.get("t") == "s":
                txt = shared[int(v.text)]
            else:
                txt = v.text or ""
            cells[col] = txt
        if cells:
            cols = sorted(cells, key=lambda s: (len(s), s))
            rows.append([cells.get(c, "") for c in cols])
    return rows

def read_xls(path):
    """Read the legacy binary .xls. pandas handles this format; the two newer
    files do not need it."""
    import pandas as pd
    x = pd.ExcelFile(path)
    d = x.parse(x.sheet_names[0], header=None, dtype=str)
    return [["" if (v is None or str(v) == "nan") else str(v) for v in row]
            for row in d.values.tolist()]

def main():
    if not os.path.isdir(os.path.dirname(OUT_DIR) or "."):
        sys.exit("Run this from the repo root, or set NA_PANEL_DATA_DIR.")
    os.makedirs(OUT_DIR, exist_ok=True)

    edges = []
    for vintage, fname in FILES:
        path = os.path.join(OUT_DIR, fname)
        if not os.path.exists(path):
            url = BASE + fname
            print("downloading", url)
            with urllib.request.urlopen(url, timeout=180) as r, \
                    open(path + ".part", "wb") as f:
                f.write(r.read())
            os.replace(path + ".part", path)
        else:
            print("using cached", path)

        rows = read_xls(path) if fname.endswith(".xls") else read_xlsx(path)

        start = None
        for i, row in enumerate(rows[:10]):
            if row and row[0].strip().endswith("NAICS Code"):
                start = i + 1
                break
        if start is None:
            sys.exit("Could not find the header row in %s - layout changed." % fname)

        n_before = len(edges)
        for row in rows[start:]:
            if len(row) < 3:
                continue
            old, new = row[0].strip(), row[2].strip()
            if not (re.fullmatch(r"\d{6}", old) and re.fullmatch(r"\d{6}", new)):
                continue
            edges.append({
                "vintage": vintage,
                "old_code": old,
                "old_title": row[1].strip().replace("\n", " ") if len(row) > 1 else "",
                "new_code": new,
                "new_title": row[3].strip().replace("\n", " ") if len(row) > 3 else "",
                "code_changed": "1" if old != new else "0",
            })
        got = len(edges) - n_before
        changed = sum(1 for e in edges[n_before:] if e["code_changed"] == "1")
        print("  %s: %d six-digit pairs, %d where the code changed"
              % (vintage, got, changed))

    if not edges:
        sys.exit("No edges parsed - refusing to write an empty concordance.")

    out = os.path.join(OUT_DIR, "naics_concordance_edges.csv")
    with open(out, "w", newline="", encoding="utf-8") as f:
        w = csv.DictWriter(f, fieldnames=list(edges[0].keys()))
        w.writeheader()
        w.writerows(edges)

    adj = collections.defaultdict(set)
    for e in edges:
        if e["code_changed"] == "1":
            adj[e["old_code"]].add(e["new_code"])
            adj[e["new_code"]].add(e["old_code"])
    seen, comps = set(), 0
    for n in adj:
        if n in seen:
            continue
        comps += 1
        stack = [n]
        while stack:
            u = stack.pop()
            if u in seen:
                continue
            seen.add(u)
            stack.extend(adj[u] - seen)

    print("\nwrote %s" % out)
    print("  %d rows; %d distinct changed-code edges; %d official lineages "
          "spanning %d codes" % (len(edges), sum(len(v) for v in adj.values()) // 2,
                                 comps, len(seen)))

if __name__ == "__main__":
    main()
